### ISSUE TITLE
Waking from sleep / screensaver causes window positions to swap

### ISSUE THREAD

**Issue by chris-kahn** [koekeishiya/chunkwm#355]
It seems like every time I walk away from my computer and come back, and unlock the screen, my entire layout gets messed up. If you have 4-5 desktops laid out nicely, this is kind of annoying. Is there a limitation from mac os here?

**Comment by neverfox** [comment]
Same here.

**Comment by koekeishiya** [comment]
Do you have a mbp (laptop with built-in display) or are you using some of the different products Apple offer, that do not have a builty-in display.

Do you use a 4k display?

I have never had this problem (even since the very beginning of development on chunkwm), but I only have a mbp (retina) and a couple of 1080p displays to test with.

I know from prior experience that macOS does seem to handle monitors differently if it is a 4k monitor for example, and this could be what we are seeing here. I'm inclined to say that this is potentially a system-limitation, due to the fact that I've never experienced it in my setup.

If the people who can reproduce this issue can try to identify commonalities in their system setup, OS versions etc, we can at least try to cover areas where this does happen, and take it from there.

**Comment by chris-kahn** [comment]
Yes, I'm using a new MBP plugged into an external 4k monitor. I keep the macbook closed though and use only the external monitor.


Включено 19 марта 2018 г. в 14:02:01, Åsmund Vikane (notifications@github.com(mailto:notifications@github.com)) написал:

**Comment by neverfox** [comment]
I have a new MBP 15. I usually have it connected to an external ultrawide HD monitor via a thunderbolt 3 dock (which is connected to the monitor via DP) and I usually keep the laptop itself closed. Whenever I wake the system, the windows have all changed position on the monitor. If they were in a spiral, they're now all equal size. Also, I sometimes "grab and go" with the laptop and whenever I disconnect or reconnect the monitor, the windows are all off the screen and I'm forced to restart chunkwm to get them to reestablish themselves within the boundary of the screen. This is unfortunately a big usability issue, because as nice as the tiling is, it becomes counterproductive if it can so easily be thrown into chaos due to my daily routine. Happy to help figure this out however I can. Here's my neofetch, if that helps:

```
OS: macOS 10.13.3 17D102 x86_64
Host: MacBookPro14,3
Kernel: 17.4.0
Uptime: 17 hours, 39 mins
Packages: 183
Shell: zsh 5.4.2
Resolution: 3440x1440@2x
DE: Aqua
WM: Kwm
Terminal: kitty
CPU: Intel i7-7820HQ (8) @ 2.90GHz
GPU: Intel HD Graphics 630, Radeon Pro 560
Memory: 4324MiB / 16384MiB
```

**Comment by chris-kahn** [comment]
I'm running the same machine but I'm still on Sierra...

```
OS: macOS Sierra 10.12.6 16G1212 x86_64
Host: MacBookPro14,3
Kernel: 16.7.0
Uptime: 5 days, 5 hours, 15 mins
Packages: 98
Shell: zsh 5.4.2
Resolution: 3008x1692@2x
DE: Aqua
WM: Kwm
Terminal: kitty
CPU: Intel i7-7820HQ (8) @ 2.90GHz
GPU: Radeon Pro 560, Intel HD Graphics 630
Memory: 6825MiB / 16384MiB
```

**Comment by neverfox** [comment]
My theory is that when the external monitor goes into power-saving mode, the system essentially doesn't believe it's there anymore and is forced to move the windows to the internal display (even thought that's "off" too for being closed).

**Comment by neverfox** [comment]
Perhaps this is related, but the window state isn't preserved on a restart of chunkwm either. Maybe that's another issue though that requires its own solution.

**Comment by koekeishiya** [comment]
> Perhaps this is related, but the window state isn't preserved on a restart of chunkwm either

That is a separate issue, but could maybe somewhat be added if I get to implementing #190

**Comment by pyro2927** [comment]
Can confirm a derivative of this issue.  I use a 2016 MBP with two external monitors (lid normally closed).  When unplug the monitors and open up the laptop, it seems to be using the resolution and layout of the external displays (1920x1080), and I need to `brew service restart chunkwm` in order to get things laying out correctly again.  Same goes for closing the lid and plugging in a monitor, it uses the screen resolution of the internal display even though it's off.

It's not a dealbreaker for me, more a strong annoyance, and I'll continue to use chunk whether or not it gets resolved.
